From 7c02e548e48dfede204dd0100a24841060432624 Mon Sep 17 00:00:00 2001
From: Hanna Hawa <hannah@marvell.com>
Date: Wed, 30 May 2018 15:00:00 +0300
Subject: [PATCH 0651/1051] pci: cp110: configure Resquester-ID to Stream-ID
 mapping

When a PCI devices access the AP internal bus it is assigned a
Stream-ID (SID) that identifies it in the System MMU (MMU-500). This
SID is calculated based on the PCI Requester-ID, a 16 bit value
composed of bus[15:8] device[7:3] and function[2:0]. Since the PCIe
Stream-ID is limited to 8 bits, a translation must take place.

This patch configures the SID mapping register to perform the
following mapping:
SID[7]   = 1. fixed prefix bit
SID[6:5] = bus[1:0]
SID[4:3] = device[1:0]
SID[2:0] = function[2:0]

The prefix is added to avoid a SID = 0 for bus = device = func = 0.
SID 0 is used by several units that wish to bypass the IOMMU like
the ICU so we need to avoid that.

Change-Id: I4a7a386d75aababeb9989ce143ff097d6cf0a434
Signed-off-by: Yehuda Yitschak <yehuday@marvell.com>
Signed-off-by: Hanna Hawa <hannah@marvell.com>
[Kevin: The original patch got from Marvell sdk10.0_19.06]
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 drivers/pci/controller/dwc/pcie-armada8k.c | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/drivers/pci/controller/dwc/pcie-armada8k.c b/drivers/pci/controller/dwc/pcie-armada8k.c
index c01cf2b59b6b..6f0362914490 100644
--- a/drivers/pci/controller/dwc/pcie-armada8k.c
+++ b/drivers/pci/controller/dwc/pcie-armada8k.c
@@ -65,6 +65,16 @@ struct armada8k_pcie {
 #define AX_USER_DOMAIN_MASK		0x3
 #define AX_USER_DOMAIN_SHIFT		4
 
+#define PCIE_STREAM_ID			(PCIE_VENDOR_REGS_OFFSET + 0x64)
+#define STREAM_ID_BUS_BITS		2
+#define STREAM_ID_DEV_BITS		2
+#define STREAM_ID_FUNC_BITS		3
+#define STREAM_ID_PREFIX		0x80
+#define PCIE_STREAM_ID_CFG		(STREAM_ID_PREFIX << 12 | \
+					STREAM_ID_BUS_BITS << 8 | \
+					STREAM_ID_DEV_BITS << 4 | \
+					STREAM_ID_FUNC_BITS)
+
 #define to_armada8k_pcie(x)	dev_get_drvdata((x)->dev)
 
 static int armada8k_pcie_link_up(struct dw_pcie *pci)
@@ -86,6 +96,9 @@ static void armada8k_pcie_establish_link(struct armada8k_pcie *pcie)
 	struct dw_pcie *pci = pcie->pci;
 	u32 reg;
 
+	/* Setup Requester-ID to Stream-ID mapping */
+	dw_pcie_writel_dbi(pci, PCIE_STREAM_ID, PCIE_STREAM_ID_CFG);
+
 	if (!dw_pcie_link_up(pci)) {
 		/* Disable LTSSM state machine to enable configuration */
 		reg = dw_pcie_readl_dbi(pci, PCIE_GLOBAL_CONTROL_REG);
-- 
2.17.1

